src/helper.cc

```diff
@@ -891,38 +890,30 @@ Helper::Client::handleFewerServers(const bool madeProgress)
 }
 
 void
-Helper::Session::HelperServerClosed(Session * const srv)
+Helper::Client::sessionClosed(SessionBase &session)
 {
-    const auto hlp = srv->parent;
-
     bool needsNewServers = false;
-    hlp->handleKilledServer(srv, needsNewServers);
+    handleKilledServer(&session, needsNewServers);
     if (needsNewServers) {
         debugs(80, DBG_IMPORTANT, "Starting new helpers");
-        helperOpenServers(hlp);
+        openSessions();
     }
+}
 
+void
+Helper::Session::HelperServerClosed(Session * const srv)
+{
+    srv->parent->sessionClosed(*srv);
     srv->dropQueued();
-
     delete srv;
 }
 
-// XXX: Almost duplicates Helper::Session::HelperServerClosed() because helperOpenServers() is not a virtual method of the `Helper::Client` class
-// TODO: Fix the `Helper::Client` class hierarchy to use virtual functions.
+// XXX: Essentially duplicates Helper::Session::HelperServerClosed() until we add SessionBase::helper().
 void
```

Comment:
This PR removes the previously disclosed blocker for removing this code duplication (i.e. the old XXX removed above). Unfortunately, we have discovered (the hard way!) another blocker (i.e. this new XXX). The new blocker is difficult to spot because both HelperServerClosed() method bodies have identical "spelling" now, but there is no `srv->helper` variable in the common base of the two `srv` classes (i.e. there is no SessionBase::helper data member).

Please do not ask to address this XXX in this minimal PR. I have checked, and adding helper() or a similar API will trigger many out-of-scope changes, some of which may be controversial. Let's continue to improve this messy helper code with small (but correct) steps.